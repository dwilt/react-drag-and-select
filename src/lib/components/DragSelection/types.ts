export interface Box {
  left: number;
  top: number;
  width: number;
  height: number;
}

export interface SelectionBox extends Box {}
