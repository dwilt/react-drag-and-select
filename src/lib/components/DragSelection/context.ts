import { createContext } from "react";
import { SelectionBox } from "./types";

export const DragSelectionContext = createContext<{
  selectionBox: SelectionBox | null;
}>({
  selectionBox: null
});
