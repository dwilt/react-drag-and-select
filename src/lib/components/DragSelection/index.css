.selection-border {
  background: rgba(0, 162, 255, 0.4);
  position: absolute;
  z-index: 99;
}
