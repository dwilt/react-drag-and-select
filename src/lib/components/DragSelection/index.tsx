import React, { useState, useEffect } from "react";
import "./index.css";
import { DragSelectionContext } from "./context";
import { SelectionBox } from "./types";

interface Point {
  x: number;
  y: number;
}

const DragSelection: React.FC = ({ children }) => {
  const [startPoint, setStartPoint] = useState<null | Point>(null);
  const [endPoint, setEndPoint] = useState<null | Point>(null);
  const [selectionBox, setSelectionBox] = useState<null | SelectionBox>(null);

  const onMouseMove = (e: MouseEvent) => {
    setEndPoint({
      x: e.pageX,
      y: e.pageY
    });
  };

  const onMouseDown = (e: React.MouseEvent) => {
    setStartPoint({
      x: e.pageX,
      y: e.pageY
    });

    document.addEventListener("mousemove", onMouseMove);
    document.addEventListener("mouseup", onMouseUp);
  };

  const onMouseUp = () => {
    document.removeEventListener("mousemove", onMouseMove);
    document.removeEventListener("mouseup", onMouseUp);

    setEndPoint(null);
    setStartPoint(null);
  };

  useEffect(() => {
    if (startPoint && endPoint) {
      setSelectionBox(
        calculateSelectionBox({
          startPoint,
          endPoint
        })
      );
    } else {
      setSelectionBox(null);
    }
  }, [endPoint, startPoint]);

  const calculateSelectionBox = ({
    startPoint,
    endPoint
  }: {
    startPoint: Point;
    endPoint: Point;
  }) => ({
    left: Math.min(startPoint.x, endPoint.x),
    top: Math.min(startPoint.y, endPoint.y),
    width: Math.abs(startPoint.x - endPoint.x),
    height: Math.abs(startPoint.y - endPoint.y)
  });

  return (
    <div
      style={{
        flex: 1,
        backgroundColor: "gray"
      }}
      {...{ onMouseDown }}
    >
      <DragSelectionContext.Provider
        value={{
          selectionBox
        }}
      >
        {children}
      </DragSelectionContext.Provider>

      {startPoint && endPoint && (
        <div className="selection-border" style={selectionBox || {}} />
      )}
    </div>
  );
};

export default DragSelection;
