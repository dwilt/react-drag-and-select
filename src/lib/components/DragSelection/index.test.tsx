import React from "react";
import ReactDOM from "react-dom";
import DragSelection from ".";
import renderer from "react-test-renderer";

it("renders without crashing", () => {
  const div = document.createElement("div");
  ReactDOM.render(<DragSelection />, div);
  ReactDOM.unmountComponentAtNode(div);
});

it("works", () => {
  const component = renderer.create(<DragSelection />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
