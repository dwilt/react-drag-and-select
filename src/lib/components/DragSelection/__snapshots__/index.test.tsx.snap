// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`works 1`] = `
<div
  onMouseDown={[Function]}
  style={
    Object {
      "backgroundColor": "gray",
      "flex": 1,
    }
  }
/>
`;
